服务端抛业务异常的时候会意外地打异常栈

目前motan服务端抛常规的业务异常的时候，会意外地打一条错误日志，并打相应的异常栈。比如我有如下一个接口，提供RPC服务给其它人使用，它的功能对指定的人打招呼，逻辑如下：

1. 所以如果参数`name`是`"Tom"`，就抛一个`IllegalArgumentException`（可能是跟Tom吵架了，所以拒绝打招呼）。
2. 对其它所有人都返回一个`"hello, " + name`的字符串，

```java
public interface GreetService {
    String sayHello(String name) throws IllegalArgumentException;
}
```

现在client调用该方法的时候传了`Tom`参数，就会在服务端打了一个类似如下的异常栈：

```
error - Exception caught when during method invocation. request:GreetService.sayHello(java.lang.String) requestId=152964379653438725
java.lang.reflect.InvocationTargetException: null
    at sun.reflect.GeneratedMethodAccessor272.invoke(Unknown Source)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:498)
    at com.weibo.api.motan.rpc.DefaultProvider.invoke(DefaultProvider.java:64)
    at com.weibo.api.motan.rpc.AbstractProvider.call(AbstractProvider.java:49)
    at com.weibo.api.motan.filter.AccessLogFilter.filter(AccessLogFilter.java:56)
    at com.weibo.api.motan.protocol.support.ProtocolFilterDecorator$2.call(ProtocolFilterDecorator.java:149)
    at com.weibo.api.motan.weibo.filter.WeiboAccessStatisticFilter.filter(WeiboAccessStatisticFilter.java:110)
    at com.weibo.api.motan.protocol.support.ProtocolFilterDecorator$2.call(ProtocolFilterDecorator.java:149)
    at com.weibo.api.motan.protocol.support.ProtocolFilterDecorator$2.call(ProtocolFilterDecorator.java:149)
    at com.weibo.api.motan.protocol.support.ProtocolFilterDecorator$2.call(ProtocolFilterDecorator.java:149)
    at com.weibo.api.motan.weibo.filter.WeiboContextFilter.filter(WeiboContextFilter.java:56)
    at com.weibo.api.motan.protocol.support.ProtocolFilterDecorator$2.call(ProtocolFilterDecorator.java:149)
    at com.weibo.api.motan.transport.ProviderMessageRouter.call(ProviderMessageRouter.java:98)
    at com.weibo.api.motan.transport.ProviderProtectedMessageRouter.call(ProviderProtectedMessageRouter.java:79)
    at com.weibo.api.motan.transport.ProviderMessageRouter.handle(ProviderMessageRouter.java:93)
    at com.weibo.api.motan.transport.support.DefaultRpcHeartbeatFactory$HeartMessageHandleWrapper.handle(DefaultRpcHeartbeatFactory.java:101)
    at com.weibo.api.motan.transport.netty.NettyChannelHandler.processRequest(NettyChannelHandler.java:137)
    at com.weibo.api.motan.transport.netty.NettyChannelHandler.access$000(NettyChannelHandler.java:43)
    at com.weibo.api.motan.transport.netty.NettyChannelHandler$1.run(NettyChannelHandler.java:112)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalArgumentException: Tom
    at GreetService.sayHello(JerryGreetService.java:32)
    ... 25 more
```

而实际上，这是一个合理的业务异常，日志应该打`INFO`级别或者`DEBUG`级别，而不应该打`ERROR`级别，更不适合打一个很长的异常栈。（想像一下一个有几万的QPS的服务处理的都是这类请求，这个服务将会被打日志打死）

经定位，这个问题是由 #501 这一个PR引入的，问题代码是

```java
try {
    Object value = method.invoke(proxyImpl, request.getArguments());
    response.setValue(value);
} catch (Exception e) {
    if (e.getCause() != null) {
        response.setException(new MotanBizException("provider call process error", e.getCause()));
    } else {
        response.setException(new MotanBizException("provider call process error", e));
    }
    // 问题在这儿，它对所有异常都打了ERROR日志
    LoggerUtil.error("Exception caught when during method invocation. request:" + request.toString(), e);
} catch (Throwable t) {
    if (t.getCause() != null) {
        response.setException(new MotanServiceException("provider has encountered a fatal error!", t.getCause()));
    } else {
        response.setException(new MotanServiceException("provider has encountered a fatal error!", t));
    }
    LoggerUtil.error("Exception caught when during method invocation. request:" + request.toString(), t);
}
```

--- a/motan-core/src/main/java/com/weibo/api/motan/rpc/DefaultProvider.java
+++ b/motan-core/src/main/java/com/weibo/api/motan/rpc/DefaultProvider.java
@@ -59,6 +59,7 @@ public class DefaultProvider<T> extends AbstractProvider<T> {
             return response;
         }
 
+        boolean defaultThrowExceptionStack = URLParamType.transExceptionStack.getBooleanValue();
         try {
             Object value = method.invoke(proxyImpl, request.getArguments());
             response.setValue(value);
@@ -69,18 +70,19 @@ public class DefaultProvider<T> extends AbstractProvider<T> {
                 response.setException(new MotanBizException("provider call process error", e));
             }
 
-            // not print detail error when exception declared in method
+            // not print stack in error log when exception declared in method
             boolean logException = true;
             for (Class<?> clazz : method.getExceptionTypes()) {
-                if (e.getCause().getClass() == clazz) {
+                if (clazz.isInstance(response.getException().getCause())) {
                     logException = false;
+                    defaultThrowExceptionStack = false;
                     break;
                 }
             }
             if (logException) {
                 LoggerUtil.error("Exception caught when during method invocation. request:" + request.toString(), e);
             } else {
-                LoggerUtil.info("Exception caught when during method invocation. request:" + request.toString());
+                LoggerUtil.info("Exception caught when during method invocation. request:" + request.toString() + ", exception:" + response.getException().getCause().toString());
             }
         } catch (Throwable t) {
             // 如果服务发生Error，将Error转化为Exception，防止拖垮调用方
@@ -95,7 +97,7 @@ public class DefaultProvider<T> extends AbstractProvider<T> {
 
         if (response.getException() != null) {
             //是否传输业务异常栈
-            boolean transExceptionStack = this.url.getBooleanParameter(URLParamType.transExceptionStack.getName(), URLParamType.transExceptionStack.getBooleanValue());
+            boolean transExceptionStack = this.url.getBooleanParameter(URLParamType.transExceptionStack.getName(), defaultThrowExceptionStack);
             if (!transExceptionStack) {//不传输业务异常栈
                 ExceptionUtil.setMockStackTrace(response.getException().getCause());
             }

--- a/motan-core/src/main/java/com/weibo/api/motan/rpc/DefaultProvider.java
+++ b/motan-core/src/main/java/com/weibo/api/motan/rpc/DefaultProvider.java
@@ -29,7 +29,6 @@ import java.lang.reflect.Method;
 /**
  * @author maijunsheng
  * @version 创建时间：2013-5-23
- *
  */
 @SpiMeta(name = "motan")
 public class DefaultProvider<T> extends AbstractProvider<T> {
@@ -41,8 +40,8 @@ public class DefaultProvider<T> extends AbstractProvider<T> {
     }
 
     @Override
-    public T getImpl(){
-    	return proxyImpl;
+    public T getImpl() {
+        return proxyImpl;
     }
 
     @Override
@@ -69,8 +68,20 @@ public class DefaultProvider<T> extends AbstractProvider<T> {
             } else {
                 response.setException(new MotanBizException("provider call process error", e));
             }
-            //服务发生错误时，显示详细日志
-            LoggerUtil.error("Exception caught when during method invocation. request:" + request.toString(), e);
+
+            // not print detail error when exception declared in method
+            boolean logException = true;
+            for (Class<?> clazz : method.getExceptionTypes()) {
+                if (e.getCause().getClass() == clazz) {
+                    logException = false;
+                    break;
+                }
+            }
+            if (logException) {
+                LoggerUtil.error("Exception caught when during method invocation. request:" + request.toString(), e);
+            } else {
+                LoggerUtil.info("Exception caught when during method invocation. request:" + request.toString());
+            }
         } catch (Throwable t) {
             // 如果服务发生Error，将Error转化为Exception，防止拖垮调用方
             if (t.getCause() != null) {

--- a/motan-demo/motan-demo-api/src/main/java/com/weibo/motan/demo/service/MotanDemoService.java
+++ b/motan-demo/motan-demo-api/src/main/java/com/weibo/motan/demo/service/MotanDemoService.java
@@ -21,8 +21,8 @@ import com.weibo.motan.demo.service.model.User;
 
 @MotanAsync
 public interface MotanDemoService {
-	String hello(String name);
+    String hello(String name);
 
-	User rename(User user, String name);
+    User rename(User user, String name) throws NullPointerException;
 
 }

--- a/motan-demo/motan-demo-api/src/main/java/com/weibo/motan/demo/service/MotanDemoService.java
+++ b/motan-demo/motan-demo-api/src/main/java/com/weibo/motan/demo/service/MotanDemoService.java
@@ -23,6 +23,6 @@ import com.weibo.motan.demo.service.model.User;
 public interface MotanDemoService {
     String hello(String name);
 
-    User rename(User user, String name) throws NullPointerException;
+    User rename(User user, String name) throws Exception;
 
 }

--- a/motan-demo/motan-demo-server/src/main/java/com/weibo/motan/demo/server/MotanDemoServiceImpl.java
+++ b/motan-demo/motan-demo-server/src/main/java/com/weibo/motan/demo/server/MotanDemoServiceImpl.java
@@ -32,7 +32,7 @@ public class MotanDemoServiceImpl implements MotanDemoService {
     }
 
     @Override
-    public User rename(User user, String name) {
+    public User rename(User user, String name) throws Exception {
         Objects.requireNonNull(user);
         System.out.println(user.getId() + " rename " + user.getName() + " to " + name);
         user.setName(name);

--- a/motan-demo/motan-demo-server/src/main/java/com/weibo/motan/demo/server/MotanDemoServiceImpl.java
+++ b/motan-demo/motan-demo-server/src/main/java/com/weibo/motan/demo/server/MotanDemoServiceImpl.java
@@ -20,6 +20,8 @@ import com.weibo.api.motan.config.springsupport.annotation.MotanService;
 import com.weibo.motan.demo.service.MotanDemoService;
 import com.weibo.motan.demo.service.model.User;
 
+import java.util.Objects;
+
 @MotanService(export = "demoMotan:8002")
 public class MotanDemoServiceImpl implements MotanDemoService {
 
@@ -31,10 +33,7 @@ public class MotanDemoServiceImpl implements MotanDemoService {
 
     @Override
     public User rename(User user, String name) {
-        if (user == null) {
-            System.out.println("user: null");
-            return null;
-        }
+        Objects.requireNonNull(user);
         System.out.println(user.getId() + " rename " + user.getName() + " to " + name);
         user.setName(name);
         return user;

